presto-main/src/main/java/com/facebook/presto/execution/TaskThresholdMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,128 @@
+package com.facebook.presto.execution;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.memory.QueryContext;
+import com.facebook.presto.memory.VoidTraversingQueryContextVisitor;
+import com.facebook.presto.operator.OperatorContext;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.annotation.Nullable;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.Collection;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class TaskThresholdMemoryRevokingScheduler
```

Comment:
I just realized memory revoking scheduler (and `MemoryRevokingScheduler`) doesn't implement any interface. Curious how are they get called? I assume it's related to the `@PostConstruct` and `@PreDestroy` annotation?

Reply:
Yup you got it. They run as a separate thread and are created at injection time. On creation (`@PostConstruct`), we register a periodically executing method to check if revoking is necessary (based on whichever spilling strategy we are using). As for which scheduler is created, that is determined in ServerMainModule, using `installModuleIf`. I was wondering too if it's worth having an interface but figured there wasn't enough reason to abstract it *yet*.

Reply:
@sachdevs : I see. Is this a mechanism from Guice? (e.g. automatically call methods annotated with `@PostConstruct`)

Reply:
Javax annotation: https://docs.oracle.com/javaee/7/api/javax/annotation/PostConstruct.html.